python: tp_vectorcall, new python struct member intro-ed in python 3.8

leads to warning, we should init for python 3.8 or newer

```
/Users/harrison37/Work/github/llnl/conduit/src/libs/conduit/python/conduit_python.cpp:2338:1: warning: missing field 'tp_vectorcall' initializer
      [-Wmissing-field-initializers]
};

```

--- a/src/libs/conduit/python/conduit_python.cpp
+++ b/src/libs/conduit/python/conduit_python.cpp
@@ -2355,7 +2355,7 @@ PyConduit_DataType_to_string(PyConduit_DataType* self,
                                      pad,
                                      eoe);
     }
-    catch(conduit::Error e)
+    catch(conduit::Error &e)
     {
         PyErr_SetString(PyExc_IOError,
                         e.message().c_str());
@@ -2420,7 +2420,7 @@ PyConduit_DataType_to_json(PyConduit_DataType* self,
                                    pad,
                                    eoe);
     }
-    catch(conduit::Error e)
+    catch(conduit::Error &e)
     {
         PyErr_SetString(PyExc_IOError,
                         e.message().c_str());
@@ -2485,7 +2485,7 @@ PyConduit_DataType_to_yaml(PyConduit_DataType* self,
                                    pad,
                                    eoe);
     }
-    catch(conduit::Error e)
+    catch(conduit::Error &e)
     {
         PyErr_SetString(PyExc_IOError,
                         e.message().c_str());
@@ -3836,7 +3836,7 @@ PyConduit_Schema_add_child(PyConduit_Schema *self,
         retval = PyConduit_Schema_Python_Wrap(&(*self->schema).add_child(std::string(name)),
                                               0); // schema owns
     }
-    catch(conduit::Error e)
+    catch(conduit::Error &e)
     {
         PyErr_SetString(PyExc_Exception,
                         e.message().c_str());
@@ -3887,7 +3887,7 @@ PyConduit_Schema_child(PyConduit_Schema *self,
             return NULL;
         }
     }
-    catch(conduit::Error e)
+    catch(conduit::Error &e)
     {
         PyErr_SetString(PyExc_Exception,
                         e.message().c_str());
@@ -3925,7 +3925,7 @@ PyConduit_Schema_rename_child(PyConduit_Schema *self,
                                    std::string(new_name));
 
     }
-    catch(conduit::Error e)
+    catch(conduit::Error &e)
     {
         PyErr_SetString(PyExc_Exception,
                         e.message().c_str());
@@ -3958,7 +3958,7 @@ PyConduit_Schema_remove_child(PyConduit_Schema *self,
     {
         self->schema->remove_child(std::string(name));
     }
-    catch(conduit::Error e)
+    catch(conduit::Error &e)
     {
         PyErr_SetString(PyExc_Exception,
                         e.message().c_str());
@@ -4005,7 +4005,7 @@ PyConduit_Schema_remove(PyConduit_Schema *self,
             return NULL;
         }
     }
-    catch(conduit::Error e)
+    catch(conduit::Error &e)
     {
         PyErr_SetString(PyExc_Exception,
                         e.message().c_str());
@@ -4185,7 +4185,7 @@ PyConduit_Schema_to_string(PyConduit_Schema* self,
                                        pad,
                                        eoe);
     }
-    catch(conduit::Error e)
+    catch(conduit::Error &e)
     {
         PyErr_SetString(PyExc_IOError,
                         e.message().c_str());
@@ -4249,7 +4249,7 @@ PyConduit_Schema_to_json(PyConduit_Schema* self,
                                      pad,
                                      eoe);
     }
-    catch(conduit::Error e)
+    catch(conduit::Error &e)
     {
         PyErr_SetString(PyExc_IOError,
                         e.message().c_str());
@@ -4313,7 +4313,7 @@ PyConduit_Schema_to_yaml(PyConduit_Schema* self,
                                      pad,
                                      eoe);
     }
-    catch(conduit::Error e)
+    catch(conduit::Error &e)
     {
         PyErr_SetString(PyExc_IOError,
                         e.message().c_str());
@@ -5260,7 +5260,7 @@ PyConduit_Node_parse(PyConduit_Node *self,
         self->node->parse(text_str,
                           protocol_str);
     }
-    catch(conduit::Error e)
+    catch(conduit::Error &e)
     {
             PyErr_SetString(PyExc_IOError,
                             e.message().c_str());
@@ -5308,7 +5308,7 @@ PyConduit_Node_save(PyConduit_Node *self,
     {
         self->node->save(path_str,protocol_str);
     }
-    catch(conduit::Error e)
+    catch(conduit::Error &e)
     {
         PyErr_SetString(PyExc_IOError,
                         e.message().c_str());
@@ -5368,7 +5368,7 @@ PyConduit_Node_load(PyConduit_Node *self,
             self->node->load(path_str,
                              *schema_ptr);
         }
-        catch(conduit::Error e)
+        catch(conduit::Error &e)
         {
             PyErr_SetString(PyExc_IOError,
                             e.message().c_str());
@@ -5390,7 +5390,7 @@ PyConduit_Node_load(PyConduit_Node *self,
             self->node->load(path_str,
                              protocol_str);
         }
-        catch(conduit::Error e)
+        catch(conduit::Error &e)
         {
             PyErr_SetString(PyExc_IOError,
                             e.message().c_str());
@@ -5443,7 +5443,7 @@ PyConduit_Node_fetch(PyConduit_Node* self,
         retval = PyConduit_Node_Python_Wrap(&(*self->node).fetch(path),
                                             0); // node owns
     }
-    catch(conduit::Error e)
+    catch(conduit::Error &e)
     {
         PyErr_SetString(PyExc_Exception,
                         e.message().c_str());
@@ -5470,7 +5470,7 @@ PyConduit_Node_fetch_existing(PyConduit_Node* self,
         retval = PyConduit_Node_Python_Wrap(&(*self->node).fetch_existing(path),
                                             0); // node owns
     }
-    catch(conduit::Error e)
+    catch(conduit::Error &e)
     {
         PyErr_SetString(PyExc_Exception,
                         e.message().c_str());
@@ -5506,7 +5506,7 @@ PyConduit_Node_add_child(PyConduit_Node *self,
         retval = PyConduit_Node_Python_Wrap(&(*self->node).add_child(std::string(name)),
                                             0); // node owns
     }
-    catch(conduit::Error e)
+    catch(conduit::Error &e)
     {
         PyErr_SetString(PyExc_Exception,
                         e.message().c_str());
@@ -5557,7 +5557,7 @@ PyConduit_Node_child(PyConduit_Node *self,
             return NULL;
         }
     }
-    catch(conduit::Error e)
+    catch(conduit::Error &e)
     {
         PyErr_SetString(PyExc_Exception,
                         e.message().c_str());
@@ -5688,7 +5688,7 @@ PyConduit_Node_remove(PyConduit_Node *self,
             return NULL;
         }
     }
-    catch(conduit::Error e)
+    catch(conduit::Error &e)
     {
         PyErr_SetString(PyExc_Exception,
                         e.message().c_str());
@@ -5721,7 +5721,7 @@ PyConduit_Node_remove_child(PyConduit_Node *self,
     {
         self->node->remove_child(std::string(name));
     }
-    catch(conduit::Error e)
+    catch(conduit::Error &e)
     {
         PyErr_SetString(PyExc_Exception,
                         e.message().c_str());
@@ -5758,7 +5758,7 @@ PyConduit_Node_rename_child(PyConduit_Node *self,
                                  std::string(new_name));
 
     }
-    catch(conduit::Error e)
+    catch(conduit::Error &e)
     {
         PyErr_SetString(PyExc_Exception,
                         e.message().c_str());
@@ -6443,7 +6443,7 @@ PyConduit_Node_to_string(PyConduit_Node* self,
                                      pad,
                                      eoe);
     }
-    catch(conduit::Error e)
+    catch(conduit::Error &e)
     {
         PyErr_SetString(PyExc_IOError,
                         e.message().c_str());
@@ -6496,7 +6496,7 @@ PyConduit_Node_to_summary_string(PyConduit_Node* self,
                                              *opts_ptr);
 
     }
-    catch(conduit::Error e)
+    catch(conduit::Error &e)
     {
         PyErr_SetString(PyExc_IOError,
                         e.message().c_str());
@@ -6570,7 +6570,7 @@ PyConduit_Node_to_json(PyConduit_Node* self,
                                    pad,
                                    eoe);
     }
-    catch(conduit::Error e)
+    catch(conduit::Error &e)
     {
         PyErr_SetString(PyExc_IOError,
                         e.message().c_str());
@@ -6645,7 +6645,7 @@ PyConduit_Node_to_yaml(PyConduit_Node* self,
                                    pad,
                                    eoe);
     }
-    catch(conduit::Error e)
+    catch(conduit::Error &e)
     {
         PyErr_SetString(PyExc_IOError,
                         e.message().c_str());

--- a/src/libs/conduit/python/conduit_python.cpp
+++ b/src/libs/conduit/python/conduit_python.cpp
@@ -32,11 +32,20 @@
 // if you do define it, you get a deprecated warning :-)
 // we suppress the deprecated warning only in 3.8.
 // 
-#define PUSH_DEP_DECL \
+
+#if PY_VERSION_HEX >= 0x03080000 && PY_VERSION_HEX < 0x03090000
+#define PRAGMA_PUSH_DEP_DECL \
      _Pragma("GCC diagnostic push") \
      _Pragma("GCC diagnostic ignored \"-Wdeprecated-declarations\"")
+#else
+#define PRAGMA_PUSH_DEP_DECL
+#endif
 
-#define POP_DEC_DECL _Pragma("GCC diagnostic pop") 
+#if PY_VERSION_HEX >= 0x03080000 && PY_VERSION_HEX < 0x03090000
+#define PRAGMA_POP_DEP_DECL _Pragma("GCC diagnostic pop")
+#else
+#define PRAGMA_POP_DEP_DECL
+#endif
 
 
 #ifdef Py_TPFLAGS_HAVE_FINALIZE
@@ -45,10 +54,7 @@
     #if PY_VERSION_HEX >= 0x03080000
         #if PY_VERSION_HEX < 0x03090000
              // python 3.8 tail
-            #define PyVarObject_TAIL \
-                 PUSH_DEP_DECL \
-                 ,0, 0, 0 \
-                 POP_DEC_DECL
+            #define PyVarObject_TAIL ,0, 0, 0 
         #else
             // python 3.9 and newer tail
             #define PyVarObject_TAIL ,0, 0
@@ -3166,8 +3172,13 @@ static PyMethodDef PyConduit_DataType_METHODS[] = {
     {NULL, NULL, 0, NULL}
 };
 
+
+
 //---------------------------------------------------------------------------//
 //---------------------------------------------------------------------------//
+
+PRAGMA_PUSH_DEP_DECL
+
 static PyTypeObject PyConduit_DataType_TYPE = {
    PyVarObject_HEAD_INIT(NULL, 0)
    "DataType",
@@ -3220,6 +3231,8 @@ static PyTypeObject PyConduit_DataType_TYPE = {
    0  /* tp_version_tag */
    PyVarObject_TAIL
 };
+
+PRAGMA_POP_DEP_DECL
 
 //---------------------------------------------------------------------------//
 static PyConduit_DataType *
@@ -3438,6 +3451,9 @@ static PyMethodDef PyConduit_Generator_METHODS[] = {
 
 //---------------------------------------------------------------------------//
 //---------------------------------------------------------------------------//
+
+PRAGMA_PUSH_DEP_DECL
+
 static PyTypeObject PyConduit_Generator_TYPE = {
    PyVarObject_HEAD_INIT(NULL, 0)
    "Generator",
@@ -3491,6 +3507,7 @@ static PyTypeObject PyConduit_Generator_TYPE = {
    PyVarObject_TAIL
 };
 
+PRAGMA_POP_DEP_DECL
 
 //---------------------------------------------------------------------------//
 static int
@@ -4461,6 +4478,9 @@ static PyMappingMethods PyConduit_Schema_as_mapping = {
 
 //---------------------------------------------------------------------------//
 //---------------------------------------------------------------------------//
+
+PRAGMA_PUSH_DEP_DECL
+
 static PyTypeObject PyConduit_Schema_TYPE = {
    PyVarObject_HEAD_INIT(NULL, 0)
    "Schema",
@@ -4514,7 +4534,7 @@ static PyTypeObject PyConduit_Schema_TYPE = {
    PyVarObject_TAIL
 };
 
-
+PRAGMA_POP_DEP_DECL
 
 //---------------------------------------------------------------------------//
 static PyObject *
@@ -4821,6 +4841,10 @@ static PyMethodDef PyConduit_NodeIterator_METHODS[] = {
 };
 
 //---------------------------------------------------------------------------//
+//---------------------------------------------------------------------------//
+
+PRAGMA_PUSH_DEP_DECL
+
 static PyTypeObject PyConduit_NodeIterator_TYPE = {
    PyVarObject_HEAD_INIT(NULL, 0)
    "Iterator",
@@ -4873,6 +4897,8 @@ static PyTypeObject PyConduit_NodeIterator_TYPE = {
    0  /* tp_version_tag */
    PyVarObject_TAIL
 };
+
+PRAGMA_POP_DEP_DECL
 
 //---------------------------------------------------------------------------//
 static PyConduit_NodeIterator *
@@ -6960,6 +6986,10 @@ static PyMappingMethods node_as_mapping = {
 };
 
 //---------------------------------------------------------------------------//
+//---------------------------------------------------------------------------//
+
+PRAGMA_PUSH_DEP_DECL
+
 static PyTypeObject PyConduit_Node_TYPE = {
    PyVarObject_HEAD_INIT(NULL, 0)
    "Node",
@@ -7013,6 +7043,7 @@ static PyTypeObject PyConduit_Node_TYPE = {
    PyVarObject_TAIL
 };
 
+PRAGMA_POP_DEP_DECL
 
 //---------------------------------------------------------------------------//
 // conduit:::about
@@ -7893,6 +7924,9 @@ static PyMethodDef PyConduit_Endianness_METHODS[] =
 
 //---------------------------------------------------------------------------//
 //---------------------------------------------------------------------------//
+
+PRAGMA_PUSH_DEP_DECL
+
 static PyTypeObject PyConduit_Endianness_TYPE = {
    PyVarObject_HEAD_INIT(NULL, 0)
    "Endianness",
@@ -7945,6 +7979,8 @@ static PyTypeObject PyConduit_Endianness_TYPE = {
    0  /* tp_version_tag */
    PyVarObject_TAIL
 };
+
+PRAGMA_POP_DEP_DECL
 
 //---------------------------------------------------------------------------//
 //---------------------------------------------------------------------------//

--- a/src/libs/conduit/python/conduit_python.cpp
+++ b/src/libs/conduit/python/conduit_python.cpp
@@ -26,9 +26,39 @@
 // to make sure we have the correct number of initializers across python
 // versions.
 //-----------------------------------------------------------------------------
+
+// helper macros for dealing with deprecated tp_print field in 
+// python 3.8. If you don't define it, you get an un-inited warning
+// if you do define it, you get a deprecated warning :-)
+// we suppress the deprecated warning only in 3.8.
+// 
+#define PUSH_DEP_DECL \
+     _Pragma("GCC diagnostic push") \
+     _Pragma("GCC diagnostic ignored \"-Wdeprecated-declarations\"")
+
+#define POP_DEC_DECL _Pragma("GCC diagnostic pop") 
+
+
 #ifdef Py_TPFLAGS_HAVE_FINALIZE
-#define PyVarObject_TAIL ,0
+    // python 3.8 adds tp_vectorcall, at end and special slot for tp_print
+    // python 3.9 removes tp_print special slot
+    #if PY_VERSION_HEX >= 0x03080000
+        #if PY_VERSION_HEX < 0x03090000
+             // python 3.8 tail
+            #define PyVarObject_TAIL \
+                 PUSH_DEP_DECL \
+                 ,0, 0, 0 \
+                 POP_DEC_DECL
+        #else
+            // python 3.9 and newer tail
+            #define PyVarObject_TAIL ,0, 0
+        #endif
+    #else
+        // python tail when finalize is part of struct
+        #define PyVarObject_TAIL ,0
+    #endif
 #else
+// python tail when finalize is not part of struct
 #define PyVarObject_TAIL
 #endif
 
@@ -3144,10 +3174,12 @@ static PyTypeObject PyConduit_DataType_TYPE = {
    sizeof(PyConduit_DataType),  /* tp_basicsize */
    0, /* tp_itemsize */
    (destructor)PyConduit_DataType_dealloc, /* tp_dealloc */
-   0, /* tp_print */
+   // tp_print was removed in Python 3.9, its now used as
+   // tp_vectorcall_offset (which we also don't use here)
+   0, /* tp_print or tp_vectorcall_offset */
    0, /* tp_getattr */
    0, /* tp_setattr */
-   0, /* tp_compare */
+   0, /* tp_compare or tp_reserved pr tp_as_async*/
    0, /* tp_repr */
    0, /* tp_as_number */
    0, /* tp_as_sequence */
@@ -3158,7 +3190,7 @@ static PyTypeObject PyConduit_DataType_TYPE = {
    0, /* getattro */
    0, /* setattro */
    0, /* asbuffer */
-   Py_TPFLAGS_DEFAULT | Py_TPFLAGS_BASETYPE,     /* flags */
+   Py_TPFLAGS_DEFAULT | Py_TPFLAGS_BASETYPE,  /* flags */
    "Conduit DataType objects",
    0, /* traverse */
    0, /* clear */
@@ -3174,18 +3206,18 @@ static PyTypeObject PyConduit_DataType_TYPE = {
    0, /* descr_get */
    0, /* gescr_set */
    0, /* dictoffset */
-   (initproc)PyConduit_DataType_init,
+   (initproc)PyConduit_DataType_init, /* tp_init */
    0, /* alloc */
-   PyConduit_DataType_new,                                   /* new */
+   PyConduit_DataType_new, /* new */
    0, /* tp_free */
    0, /* tp_is_gc */
    0, /* tp_bases */
    0, /* tp_mro */
    0, /* tp_cache */
    0, /* tp_subclasses */
-   0,  /* tp_weaklist */
-   0,
-   0
+   0, /* tp_weaklist */
+   0, /* tp_del */
+   0  /* tp_version_tag */
    PyVarObject_TAIL
 };
 
@@ -3412,7 +3444,9 @@ static PyTypeObject PyConduit_Generator_TYPE = {
    sizeof(PyConduit_Generator),  /* tp_basicsize */
    0, /* tp_itemsize */
    (destructor)PyConduit_Generator_dealloc, /* tp_dealloc */
-   0, /* tp_print */
+   // tp_print was removed in Python 3.9, its now used as
+   // tp_vectorcall_offset (which we also don't use here)
+   0, /* tp_print or tp_vectorcall_offset */
    0, /* tp_getattr */
    0, /* tp_setattr */
    0, /* tp_compare */
@@ -3442,18 +3476,18 @@ static PyTypeObject PyConduit_Generator_TYPE = {
    0, /* descr_get */
    0, /* gescr_set */
    0, /* dictoffset */
-   (initproc)PyConduit_Generator_init,
+   (initproc)PyConduit_Generator_init,  /* tp_init */
    0, /* alloc */
-   PyConduit_Generator_new,                       /* new */
+   PyConduit_Generator_new,  /* new */
    0, /* tp_free */
    0, /* tp_is_gc */
    0, /* tp_bases */
    0, /* tp_mro */
    0, /* tp_cache */
    0, /* tp_subclasses */
-   0,  /* tp_weaklist */
-   0,
-   0
+   0, /* tp_weaklist */
+   0, /* tp_del */
+   0  /* tp_version_tag */
    PyVarObject_TAIL
 };
 
@@ -4433,7 +4467,9 @@ static PyTypeObject PyConduit_Schema_TYPE = {
    sizeof(PyConduit_Schema),  /* tp_basicsize */
    0, /* tp_itemsize */
    (destructor)PyConduit_Schema_dealloc, /* tp_dealloc */
-   0, /* tp_print */
+   // tp_print was removed in Python 3.9, its now used as
+   // tp_vectorcall_offset (which we also don't use here)
+   0, /* tp_print or tp_vectorcall_offset */
    0, /* tp_getattr */
    0, /* tp_setattr */
    0, /* tp_compare */
@@ -4472,9 +4508,9 @@ static PyTypeObject PyConduit_Schema_TYPE = {
    0, /* tp_mro */
    0, /* tp_cache */
    0, /* tp_subclasses */
-   0,  /* tp_weaklist */
-   0,
-   0
+   0, /* tp_weaklist */
+   0, /* tp_del */
+   0  /* tp_version_tag */
    PyVarObject_TAIL
 };
 
@@ -4791,7 +4827,9 @@ static PyTypeObject PyConduit_NodeIterator_TYPE = {
    sizeof(PyConduit_NodeIterator),  /* tp_basicsize */
    0, /* tp_itemsize */
    (destructor)PyConduit_NodeIterator_dealloc,   /* tp_dealloc */
-   0, /* tp_print */
+   // tp_print was removed in Python 3.9, its now used as
+   // tp_vectorcall_offset (which we also don't use here)
+   0, /* tp_print or tp_vectorcall_offset */
    0, /* tp_getattr */
    0, /* tp_setattr */
    0, /* tp_compare */
@@ -4830,10 +4868,10 @@ static PyTypeObject PyConduit_NodeIterator_TYPE = {
    0, /* tp_mro */
    0, /* tp_cache */
    0, /* tp_subclasses */
-   0,  /* tp_weaklist */
-   0,
-   0
-   PyVarObject_TAIL       
+   0, /* tp_weaklist */
+   0, /* tp_del */
+   0  /* tp_version_tag */
+   PyVarObject_TAIL
 };
 
 //---------------------------------------------------------------------------//
@@ -6928,7 +6966,9 @@ static PyTypeObject PyConduit_Node_TYPE = {
    sizeof(PyConduit_Node),  /* tp_basicsize */
    0, /* tp_itemsize */
    (destructor)PyConduit_Node_dealloc,                   /* tp_dealloc */
-   0, /* tp_print */
+   // tp_print was removed in Python 3.9, its now used as
+   // tp_vectorcall_offset (which we also don't use here)
+   0, /* tp_print or tp_vectorcall_offset */
    0, /* tp_getattr */
    0, /* tp_setattr */
    0, /* tp_compare */
@@ -6967,9 +7007,9 @@ static PyTypeObject PyConduit_Node_TYPE = {
    0, /* tp_mro */
    0, /* tp_cache */
    0, /* tp_subclasses */
-   0,  /* tp_weaklist */
-   0,
-   0
+   0, /* tp_weaklist */
+   0, /* tp_del */
+   0  /* tp_version_tag */
    PyVarObject_TAIL
 };
 
@@ -7859,7 +7899,9 @@ static PyTypeObject PyConduit_Endianness_TYPE = {
    0, // sizeof(PyConduit_Endianness), /* tp_basicsize */
    0, /* tp_itemsize */
    0, /* tp_dealloc */
-   0, /* tp_print */
+   // tp_print was removed in Python 3.9, its now used as
+   // tp_vectorcall_offset (which we also don't use here)
+   0, /* tp_print or tp_vectorcall_offset */
    0, /* tp_getattr */
    0, /* tp_setattr */
    0, /* tp_compare */
@@ -7899,8 +7941,8 @@ static PyTypeObject PyConduit_Endianness_TYPE = {
    0, /* tp_cache */
    0, /* tp_subclasses */
    0,  /* tp_weaklist */
-   0,
-   0
+   0, /* tp_del */
+   0  /* tp_version_tag */
    PyVarObject_TAIL
 };
 

--- a/src/libs/conduit/python/conduit_utils_python.cpp
+++ b/src/libs/conduit/python/conduit_utils_python.cpp
@@ -126,7 +126,7 @@ PyConduit_Utils_format(PyObject *, //self
                                          (index_t)map_index);
         }
     }
-    catch(conduit::Error e)
+    catch(conduit::Error &e)
     {
         PyErr_SetString(PyExc_IOError,
                         e.message().c_str());

--- a/src/libs/relay/conduit_relay_io_blueprint.cpp
+++ b/src/libs/relay/conduit_relay_io_blueprint.cpp
@@ -1257,7 +1257,7 @@ void write_mesh(const Node &mesh,
                                 // update status, we are done with this doman
                                 local_domain_status[d] = 0;
                             }
-                            catch(conduit::Error e)
+                            catch(conduit::Error &e)
                             {
                                 local_all_is_good = 0;
                                 local_io_exception_msg = e.message();

--- a/src/libs/relay/conduit_relay_web.cpp
+++ b/src/libs/relay/conduit_relay_web.cpp
@@ -222,7 +222,7 @@ public:
                 CONDUIT_INFO("WebSocket received message:" << n.to_json());
                 // TODO: Call recv handler callback.
             }
-            catch(conduit::Error e)
+            catch(conduit::Error &e)
             {
                  CONDUIT_INFO("Error parsing JSON response from browser\n" 
                                << e.message());

--- a/src/libs/relay/python/conduit_relay_io_blueprint_python.cpp
+++ b/src/libs/relay/python/conduit_relay_io_blueprint_python.cpp
@@ -34,17 +34,6 @@
 using namespace conduit;
 using namespace conduit::relay::io;
 
-//-----------------------------------------------------------------------------
-// PyVarObject_TAIL is used at the end of each PyVarObject def
-// to make sure we have the correct number of initializers across python
-// versions.
-//-----------------------------------------------------------------------------
-#ifdef Py_TPFLAGS_HAVE_FINALIZE
-#define PyVarObject_TAIL ,0
-#else
-#define PyVarObject_TAIL
-#endif
-
 //---------------------------------------------------------------------------//
 // conduit::relay::io::blueprint::write_mesh
 //---------------------------------------------------------------------------//

--- a/src/libs/relay/python/conduit_relay_io_blueprint_python.cpp
+++ b/src/libs/relay/python/conduit_relay_io_blueprint_python.cpp
@@ -109,7 +109,7 @@ PyRelay_io_blueprint_write_mesh(PyObject *, //self
                                          protocol_str,
                                          *opts_ptr);
     }
-    catch(conduit::Error e)
+    catch(conduit::Error &e)
     {
         PyErr_SetString(PyExc_IOError,
                         e.message().c_str());
@@ -184,7 +184,7 @@ PyRelay_io_blueprint_read_mesh(PyObject *, //self
                                         *opts_ptr,
                                         node);
     }
-    catch(conduit::Error e)
+    catch(conduit::Error &e)
     {
         PyErr_SetString(PyExc_IOError,
                         e.message().c_str());

--- a/src/libs/relay/python/conduit_relay_io_python.cpp
+++ b/src/libs/relay/python/conduit_relay_io_python.cpp
@@ -47,9 +47,39 @@ using namespace conduit::relay::io;
 // to make sure we have the correct number of initializers across python
 // versions.
 //-----------------------------------------------------------------------------
+
+// helper macros for dealing with deprecated tp_print field in 
+// python 3.8. If you don't define it, you get an un-inited warning
+// if you do define it, you get a deprecated warning :-)
+// we suppress the deprecated warning only in 3.8.
+// 
+#define PUSH_DEP_DECL \
+     _Pragma("GCC diagnostic push") \
+     _Pragma("GCC diagnostic ignored \"-Wdeprecated-declarations\"")
+
+#define POP_DEC_DECL _Pragma("GCC diagnostic pop") 
+
+
 #ifdef Py_TPFLAGS_HAVE_FINALIZE
-#define PyVarObject_TAIL ,0
+    // python 3.8 adds tp_vectorcall, at end and special slot for tp_print
+    // python 3.9 removes tp_print special slot
+    #if PY_VERSION_HEX >= 0x03080000
+        #if PY_VERSION_HEX < 0x03090000
+             // python 3.8 tail
+            #define PyVarObject_TAIL \
+                 PUSH_DEP_DECL \
+                 ,0, 0, 0 \
+                 POP_DEC_DECL
+        #else
+            // python 3.9 and newer tail
+            #define PyVarObject_TAIL ,0, 0
+        #endif
+    #else
+        // python tail when finalize is part of struct
+        #define PyVarObject_TAIL ,0
+    #endif
 #else
+// python tail when finalize is not part of struct
 #define PyVarObject_TAIL
 #endif
 
@@ -520,7 +550,9 @@ static PyTypeObject PyRelay_IOHandle_TYPE = {
    sizeof(PyRelay_IOHandle),  /* tp_basicsize */
    0, /* tp_itemsize */
    (destructor)PyRelay_IOHandle_dealloc,   /* tp_dealloc */
-   0, /* tp_print */
+   // tp_print was removed in Python 3.9, its now used as
+   // tp_vectorcall_offset (which we also don't use here)
+   0, /* tp_print or tp_vectorcall_offset */
    0, /* tp_getattr */
    0, /* tp_setattr */
    0, /* tp_compare */
@@ -559,9 +591,9 @@ static PyTypeObject PyRelay_IOHandle_TYPE = {
    0, /* tp_mro */
    0, /* tp_cache */
    0, /* tp_subclasses */
-   0,  /* tp_weaklist */
-   0,
-   0
+   0, /* tp_weaklist */
+   0, /* tp_del */
+   0  /* tp_version_tag */
    PyVarObject_TAIL
 };
 

--- a/src/libs/relay/python/conduit_relay_io_python.cpp
+++ b/src/libs/relay/python/conduit_relay_io_python.cpp
@@ -224,7 +224,7 @@ PyRelay_IOHandle_open(PyRelay_IOHandle *self,
                            protocol_str,
                            *opts_ptr);
     }
-    catch(conduit::Error e)
+    catch(conduit::Error &e)
     {
         PyErr_SetString(PyExc_IOError,
                         e.message().c_str());
@@ -294,7 +294,7 @@ PyRelay_IOHandle_read(PyRelay_IOHandle *self,
             self->handle->read(std::string(path),*node_ptr);
         }
     }
-    catch(conduit::Error e)
+    catch(conduit::Error &e)
     {
         PyErr_SetString(PyExc_IOError,
                         e.message().c_str());
@@ -349,7 +349,7 @@ PyRelay_IOHandle_write(PyRelay_IOHandle *self,
             self->handle->write(*node_ptr,std::string(path));
         }
     }
-    catch(conduit::Error e)
+    catch(conduit::Error &e)
     {
         PyErr_SetString(PyExc_IOError,
                         e.message().c_str());
@@ -391,7 +391,7 @@ PyRelay_IOHandle_list_child_names(PyRelay_IOHandle *self,
                                            cld_names);
         }
     }
-    catch(conduit::Error e)
+    catch(conduit::Error &e)
     {
         PyErr_SetString(PyExc_IOError,
                         e.message().c_str());
@@ -438,7 +438,7 @@ PyRelay_IOHandle_has_path(PyRelay_IOHandle *self,
     {
         res = self->handle->has_path(std::string(path));
     }
-    catch(conduit::Error e)
+    catch(conduit::Error &e)
     {
         PyErr_SetString(PyExc_IOError,
                         e.message().c_str());
@@ -475,7 +475,7 @@ PyRelay_IOHandle_remove(PyRelay_IOHandle *self,
     {
         self->handle->remove(std::string(path));
     }
-    catch(conduit::Error e)
+    catch(conduit::Error &e)
     {
         PyErr_SetString(PyExc_IOError,
                         e.message().c_str());
@@ -493,7 +493,7 @@ PyRelay_IOHandle_close(PyRelay_IOHandle *self)
     {
         self->handle->close();
     }
-    catch(conduit::Error e)
+    catch(conduit::Error &e)
     {
         PyErr_SetString(PyExc_IOError,
                         e.message().c_str());
@@ -698,7 +698,7 @@ PyRelay_io_save(PyObject *, //self
                         protocol_str,
                         *opts_ptr);
     }
-    catch(conduit::Error e)
+    catch(conduit::Error &e)
     {
         PyErr_SetString(PyExc_IOError,
                         e.message().c_str());
@@ -783,7 +783,7 @@ PyRelay_io_save_merged(PyObject *, //self
                                protocol_str,
                                *opts_ptr);
     }
-    catch(conduit::Error e)
+    catch(conduit::Error &e)
     {
         PyErr_SetString(PyExc_IOError,
                         e.message().c_str());
@@ -843,7 +843,7 @@ PyRelay_io_load(PyObject *, //self
                         protocol_str,
                         node);
     }
-    catch(conduit::Error e)
+    catch(conduit::Error &e)
     {
         PyErr_SetString(PyExc_IOError,
                         e.message().c_str());
@@ -905,7 +905,7 @@ PyRelay_io_load_merged(PyObject *, //self
                                protocol_str,
                                node);
     }
-    catch(conduit::Error e)
+    catch(conduit::Error &e)
     {
         PyErr_SetString(PyExc_IOError,
                         e.message().c_str());

--- a/src/libs/relay/python/conduit_relay_io_python.cpp
+++ b/src/libs/relay/python/conduit_relay_io_python.cpp
@@ -53,11 +53,20 @@ using namespace conduit::relay::io;
 // if you do define it, you get a deprecated warning :-)
 // we suppress the deprecated warning only in 3.8.
 // 
-#define PUSH_DEP_DECL \
+
+#if PY_VERSION_HEX >= 0x03080000 && PY_VERSION_HEX < 0x03090000
+#define PRAGMA_PUSH_DEP_DECL \
      _Pragma("GCC diagnostic push") \
      _Pragma("GCC diagnostic ignored \"-Wdeprecated-declarations\"")
+#else
+#define PRAGMA_PUSH_DEP_DECL
+#endif
 
-#define POP_DEC_DECL _Pragma("GCC diagnostic pop") 
+#if PY_VERSION_HEX >= 0x03080000 && PY_VERSION_HEX < 0x03090000
+#define PRAGMA_POP_DEP_DECL _Pragma("GCC diagnostic pop")
+#else
+#define PRAGMA_POP_DEP_DECL
+#endif
 
 
 #ifdef Py_TPFLAGS_HAVE_FINALIZE
@@ -66,10 +75,7 @@ using namespace conduit::relay::io;
     #if PY_VERSION_HEX >= 0x03080000
         #if PY_VERSION_HEX < 0x03090000
              // python 3.8 tail
-            #define PyVarObject_TAIL \
-                 PUSH_DEP_DECL \
-                 ,0, 0, 0 \
-                 POP_DEC_DECL
+            #define PyVarObject_TAIL ,0, 0, 0 
         #else
             // python 3.9 and newer tail
             #define PyVarObject_TAIL ,0, 0
@@ -544,6 +550,9 @@ static PyMethodDef PyRelay_IOHandle_METHODS[] = {
 
 //---------------------------------------------------------------------------//
 //---------------------------------------------------------------------------//
+
+PRAGMA_PUSH_DEP_DECL
+
 static PyTypeObject PyRelay_IOHandle_TYPE = {
    PyVarObject_HEAD_INIT(NULL, 0)
    "IOHandle",
@@ -597,6 +606,7 @@ static PyTypeObject PyRelay_IOHandle_TYPE = {
    PyVarObject_TAIL
 };
 
+PRAGMA_POP_DEP_DECL
 
 //---------------------------------------------------------------------------//
 // conduit::relay::io::about

--- a/src/libs/relay/python/conduit_relay_mpi_python.cpp
+++ b/src/libs/relay/python/conduit_relay_mpi_python.cpp
@@ -267,7 +267,7 @@ PyRelay_MPI_rank(PyObject *, //self
     {
         rank = relay::mpi::rank(comm);
     }
-    catch(conduit::Error e)
+    catch(conduit::Error &e)
     {
         PyErr_SetString(PyExc_Exception,
                         e.message().c_str());
@@ -310,7 +310,7 @@ PyRelay_MPI_size(PyObject *, //self
     {
         size = relay::mpi::size(comm);
     }
-    catch(conduit::Error e)
+    catch(conduit::Error &e)
     {
         PyErr_SetString(PyExc_Exception,
                         e.message().c_str());
@@ -374,7 +374,7 @@ PyRelay_MPI_send(PyObject *, //self
     {
         relay::mpi::send(node, dest, tag, comm);
     }
-    catch(conduit::Error e)
+    catch(conduit::Error &e)
     {
         PyErr_SetString(PyExc_Exception,
                         e.message().c_str());
@@ -435,7 +435,7 @@ PyRelay_MPI_recv(PyObject *, //self
     {
         relay::mpi::recv(node, source, tag, comm);
     }
-    catch(conduit::Error e)
+    catch(conduit::Error &e)
     {
         PyErr_SetString(PyExc_Exception,
                         e.message().c_str());
@@ -495,7 +495,7 @@ PyRelay_MPI_send_using_schema(PyObject *, //self
     {
         relay::mpi::send_using_schema(node, dest, tag, comm);
     }
-    catch(conduit::Error e)
+    catch(conduit::Error &e)
     {
         PyErr_SetString(PyExc_Exception,
                         e.message().c_str());
@@ -555,7 +555,7 @@ PyRelay_MPI_recv_using_schema(PyObject *, //self
     {
         relay::mpi::recv_using_schema(node, source, tag, comm);
     }
-    catch(conduit::Error e)
+    catch(conduit::Error &e)
     {
         PyErr_SetString(PyExc_Exception,
                         e.message().c_str());
@@ -674,7 +674,7 @@ PyRelay_MPI_sum_reduce(PyObject *, //self
                                root,
                                comm);
     }
-    catch(conduit::Error e)
+    catch(conduit::Error &e)
     {
         PyErr_SetString(PyExc_Exception,
                         e.message().c_str());
@@ -749,7 +749,7 @@ PyRelay_MPI_min_reduce(PyObject *, //self
                                root,
                                comm);
     }
-    catch(conduit::Error e)
+    catch(conduit::Error &e)
     {
         PyErr_SetString(PyExc_Exception,
                         e.message().c_str());
@@ -825,7 +825,7 @@ PyRelay_MPI_max_reduce(PyObject *, //self
                                root,
                                comm);
     }
-    catch(conduit::Error e)
+    catch(conduit::Error &e)
     {
         PyErr_SetString(PyExc_Exception,
                         e.message().c_str());
@@ -901,7 +901,7 @@ PyRelay_MPI_prod_reduce(PyObject *, //self
                                root,
                                comm);
     }
-    catch(conduit::Error e)
+    catch(conduit::Error &e)
     {
         PyErr_SetString(PyExc_Exception,
                         e.message().c_str());
@@ -972,7 +972,7 @@ PyRelay_MPI_sum_all_reduce(PyObject *, //self
                                    recv_node,
                                    comm);
     }
-    catch(conduit::Error e)
+    catch(conduit::Error &e)
     {
         PyErr_SetString(PyExc_Exception,
                         e.message().c_str());
@@ -1044,7 +1044,7 @@ PyRelay_MPI_min_all_reduce(PyObject *, //self
                                    recv_node,
                                    comm);
     }
-    catch(conduit::Error e)
+    catch(conduit::Error &e)
     {
         PyErr_SetString(PyExc_Exception,
                         e.message().c_str());
@@ -1115,7 +1115,7 @@ PyRelay_MPI_max_all_reduce(PyObject *, //self
                                    recv_node,
                                    comm);
     }
-    catch(conduit::Error e)
+    catch(conduit::Error &e)
     {
         PyErr_SetString(PyExc_Exception,
                         e.message().c_str());
@@ -1186,7 +1186,7 @@ PyRelay_MPI_prod_all_reduce(PyObject *, //self
                                     recv_node,
                                     comm);
     }
-    catch(conduit::Error e)
+    catch(conduit::Error &e)
     {
         PyErr_SetString(PyExc_Exception,
                         e.message().c_str());
@@ -1272,7 +1272,7 @@ PyRelay_MPI_prod_all_reduce(PyObject *, //self
 //                           comm,
 //                           request);
 //     }
-//     catch(conduit::Error e)
+//     catch(conduit::Error &e)
 //     {
 //         PyErr_SetString(PyExc_Exception,
 //                         e.message().c_str());
@@ -1358,7 +1358,7 @@ PyRelay_MPI_prod_all_reduce(PyObject *, //self
 //                           comm,
 //                           request);
 //     }
-//     catch(conduit::Error e)
+//     catch(conduit::Error &e)
 //     {
 //         PyErr_SetString(PyExc_Exception,
 //                         e.message().c_str());
@@ -1451,7 +1451,7 @@ PyRelay_MPI_gather(PyObject *, //self
                            root,
                            comm);
     }
-    catch(conduit::Error e)
+    catch(conduit::Error &e)
     {
         PyErr_SetString(PyExc_Exception,
                         e.message().c_str());
@@ -1523,7 +1523,7 @@ PyRelay_MPI_gather_using_schema(PyObject *, //self
                                         root,
                                         comm);
     }
-    catch(conduit::Error e)
+    catch(conduit::Error &e)
     {
         PyErr_SetString(PyExc_Exception,
                         e.message().c_str());
@@ -1591,7 +1591,7 @@ PyRelay_MPI_all_gather(PyObject *, //self
                                recv_node,
                                comm);
     }
-    catch(conduit::Error e)
+    catch(conduit::Error &e)
     {
         PyErr_SetString(PyExc_Exception,
                         e.message().c_str());
@@ -1659,7 +1659,7 @@ PyRelay_MPI_all_gather_using_schema(PyObject *, //self
                                             recv_node,
                                             comm);
     }
-    catch(conduit::Error e)
+    catch(conduit::Error &e)
     {
         PyErr_SetString(PyExc_Exception,
                         e.message().c_str());
@@ -1720,7 +1720,7 @@ PyRelay_MPI_broadcast(PyObject *, //self
     {
         relay::mpi::broadcast(node, root, comm);
     }
-    catch(conduit::Error e)
+    catch(conduit::Error &e)
     {
         PyErr_SetString(PyExc_Exception,
                         e.message().c_str());
@@ -1777,7 +1777,7 @@ PyRelay_MPI_broadcast_using_schema(PyObject *, //self
     {
         relay::mpi::broadcast_using_schema(node, root, comm);
     }
-    catch(conduit::Error e)
+    catch(conduit::Error &e)
     {
         PyErr_SetString(PyExc_Exception,
                         e.message().c_str());

--- a/src/libs/relay/python/conduit_relay_mpi_python.cpp
+++ b/src/libs/relay/python/conduit_relay_mpi_python.cpp
@@ -47,9 +47,39 @@ using namespace conduit::relay::mpi;
 // to make sure we have the correct number of initializers across python
 // versions.
 //-----------------------------------------------------------------------------
+
+// helper macros for dealing with deprecated tp_print field in 
+// python 3.8. If you don't define it, you get an un-inited warning
+// if you do define it, you get a deprecated warning :-)
+// we suppress the deprecated warning only in 3.8.
+// 
+#define PUSH_DEP_DECL \
+     _Pragma("GCC diagnostic push") \
+     _Pragma("GCC diagnostic ignored \"-Wdeprecated-declarations\"")
+
+#define POP_DEC_DECL _Pragma("GCC diagnostic pop") 
+
+
 #ifdef Py_TPFLAGS_HAVE_FINALIZE
-#define PyVarObject_TAIL ,0
+    // python 3.8 adds tp_vectorcall, at end and special slot for tp_print
+    // python 3.9 removes tp_print special slot
+    #if PY_VERSION_HEX >= 0x03080000
+        #if PY_VERSION_HEX < 0x03090000
+             // python 3.8 tail
+            #define PyVarObject_TAIL \
+                 PUSH_DEP_DECL \
+                 ,0, 0, 0 \
+                 POP_DEC_DECL
+        #else
+            // python 3.9 and newer tail
+            #define PyVarObject_TAIL ,0, 0
+        #endif
+    #else
+        // python tail when finalize is part of struct
+        #define PyVarObject_TAIL ,0
+    #endif
 #else
+// python tail when finalize is not part of struct
 #define PyVarObject_TAIL
 #endif
 
@@ -121,7 +151,9 @@ static PyTypeObject PyRelay_MPI_Request_TYPE = {
    sizeof(PyRelay_MPI_Request),  /* tp_basicsize */
    0, /* tp_itemsize */
    (destructor)PyRelay_MPI_Request_dealloc, /* tp_dealloc */
-   0, /* tp_print */
+   // tp_print was removed in Python 3.9, its now used as
+   // tp_vectorcall_offset (which we also don't use here)
+   0, /* tp_print or tp_vectorcall_offset */
    0, /* tp_getattr */
    0, /* tp_setattr */
    0, /* tp_compare */
@@ -160,9 +192,9 @@ static PyTypeObject PyRelay_MPI_Request_TYPE = {
    0, /* tp_mro */
    0, /* tp_cache */
    0, /* tp_subclasses */
-   0,  /* tp_weaklist */
-   0,
-   0
+   0, /* tp_weaklist */
+   0, /* tp_del */
+   0  /* tp_version_tag */
    PyVarObject_TAIL
 };
 

--- a/src/libs/relay/python/conduit_relay_mpi_python.cpp
+++ b/src/libs/relay/python/conduit_relay_mpi_python.cpp
@@ -53,11 +53,20 @@ using namespace conduit::relay::mpi;
 // if you do define it, you get a deprecated warning :-)
 // we suppress the deprecated warning only in 3.8.
 // 
-#define PUSH_DEP_DECL \
+
+#if PY_VERSION_HEX >= 0x03080000 && PY_VERSION_HEX < 0x03090000
+#define PRAGMA_PUSH_DEP_DECL \
      _Pragma("GCC diagnostic push") \
      _Pragma("GCC diagnostic ignored \"-Wdeprecated-declarations\"")
+#else
+#define PRAGMA_PUSH_DEP_DECL
+#endif
 
-#define POP_DEC_DECL _Pragma("GCC diagnostic pop") 
+#if PY_VERSION_HEX >= 0x03080000 && PY_VERSION_HEX < 0x03090000
+#define PRAGMA_POP_DEP_DECL _Pragma("GCC diagnostic pop")
+#else
+#define PRAGMA_POP_DEP_DECL
+#endif
 
 
 #ifdef Py_TPFLAGS_HAVE_FINALIZE
@@ -66,10 +75,7 @@ using namespace conduit::relay::mpi;
     #if PY_VERSION_HEX >= 0x03080000
         #if PY_VERSION_HEX < 0x03090000
              // python 3.8 tail
-            #define PyVarObject_TAIL \
-                 PUSH_DEP_DECL \
-                 ,0, 0, 0 \
-                 POP_DEC_DECL
+            #define PyVarObject_TAIL ,0, 0, 0 
         #else
             // python 3.9 and newer tail
             #define PyVarObject_TAIL ,0, 0
@@ -82,8 +88,6 @@ using namespace conduit::relay::mpi;
 // python tail when finalize is not part of struct
 #define PyVarObject_TAIL
 #endif
-
-
 
 //---------------------------------------------------------------------------//
 struct PyRelay_MPI_Request
@@ -145,6 +149,9 @@ static PyMethodDef PyRelay_MPI_Request_METHODS[] = {
 };
 
 //---------------------------------------------------------------------------//
+
+PRAGMA_PUSH_DEP_DECL
+
 static PyTypeObject PyRelay_MPI_Request_TYPE = {
    PyVarObject_HEAD_INIT(NULL, 0)
    "Request",
@@ -197,6 +204,8 @@ static PyTypeObject PyRelay_MPI_Request_TYPE = {
    0  /* tp_version_tag */
    PyVarObject_TAIL
 };
+
+PRAGMA_POP_DEP_DECL
 
 //---------------------------------------------------------------------------//
 // TODO: Enable when async support is enabled

--- a/src/libs/relay/python/conduit_relay_web_python.cpp
+++ b/src/libs/relay/python/conduit_relay_web_python.cpp
@@ -47,9 +47,39 @@ using namespace conduit::relay::web;
 // to make sure we have the correct number of initializers across python
 // versions.
 //-----------------------------------------------------------------------------
+
+// helper macros for dealing with deprecated tp_print field in 
+// python 3.8. If you don't define it, you get an un-inited warning
+// if you do define it, you get a deprecated warning :-)
+// we suppress the deprecated warning only in 3.8.
+// 
+#define PUSH_DEP_DECL \
+     _Pragma("GCC diagnostic push") \
+     _Pragma("GCC diagnostic ignored \"-Wdeprecated-declarations\"")
+
+#define POP_DEC_DECL _Pragma("GCC diagnostic pop") 
+
+
 #ifdef Py_TPFLAGS_HAVE_FINALIZE
-#define PyVarObject_TAIL ,0
+    // python 3.8 adds tp_vectorcall, at end and special slot for tp_print
+    // python 3.9 removes tp_print special slot
+    #if PY_VERSION_HEX >= 0x03080000
+        #if PY_VERSION_HEX < 0x03090000
+             // python 3.8 tail
+            #define PyVarObject_TAIL \
+                 PUSH_DEP_DECL \
+                 ,0, 0, 0 \
+                 POP_DEC_DECL
+        #else
+            // python 3.9 and newer tail
+            #define PyVarObject_TAIL ,0, 0
+        #endif
+    #else
+        // python tail when finalize is part of struct
+        #define PyVarObject_TAIL ,0
+    #endif
 #else
+// python tail when finalize is not part of struct
 #define PyVarObject_TAIL
 #endif
 
@@ -513,7 +543,9 @@ static PyTypeObject PyRelay_Web_WebServer_TYPE = {
    sizeof(PyRelay_Web_WebServer),  /* tp_basicsize */
    0, /* tp_itemsize */
    (destructor)PyRelay_Web_WebServer_dealloc, /* tp_dealloc */
-   0, /* tp_print */
+   // tp_print was removed in Python 3.9, its now used as
+   // tp_vectorcall_offset (which we also don't use here)
+   0, /* tp_print or tp_vectorcall_offset */
    0, /* tp_getattr */
    0, /* tp_setattr */
    0, /* tp_compare */
@@ -553,8 +585,8 @@ static PyTypeObject PyRelay_Web_WebServer_TYPE = {
    0, /* tp_cache */
    0, /* tp_subclasses */
    0,  /* tp_weaklist */
-   0,
-   0
+   0, /* tp_del */
+   0  /* tp_version_tag */
    PyVarObject_TAIL
 };
 

--- a/src/libs/relay/python/conduit_relay_web_python.cpp
+++ b/src/libs/relay/python/conduit_relay_web_python.cpp
@@ -53,11 +53,20 @@ using namespace conduit::relay::web;
 // if you do define it, you get a deprecated warning :-)
 // we suppress the deprecated warning only in 3.8.
 // 
-#define PUSH_DEP_DECL \
+
+#if PY_VERSION_HEX >= 0x03080000 && PY_VERSION_HEX < 0x03090000
+#define PRAGMA_PUSH_DEP_DECL \
      _Pragma("GCC diagnostic push") \
      _Pragma("GCC diagnostic ignored \"-Wdeprecated-declarations\"")
+#else
+#define PRAGMA_PUSH_DEP_DECL
+#endif
 
-#define POP_DEC_DECL _Pragma("GCC diagnostic pop") 
+#if PY_VERSION_HEX >= 0x03080000 && PY_VERSION_HEX < 0x03090000
+#define PRAGMA_POP_DEP_DECL _Pragma("GCC diagnostic pop")
+#else
+#define PRAGMA_POP_DEP_DECL
+#endif
 
 
 #ifdef Py_TPFLAGS_HAVE_FINALIZE
@@ -66,10 +75,7 @@ using namespace conduit::relay::web;
     #if PY_VERSION_HEX >= 0x03080000
         #if PY_VERSION_HEX < 0x03090000
              // python 3.8 tail
-            #define PyVarObject_TAIL \
-                 PUSH_DEP_DECL \
-                 ,0, 0, 0 \
-                 POP_DEC_DECL
+            #define PyVarObject_TAIL ,0, 0, 0 
         #else
             // python 3.9 and newer tail
             #define PyVarObject_TAIL ,0, 0
@@ -537,6 +543,9 @@ static PyMethodDef PyRelay_Web_WebServer_METHODS[] = {
 
 //---------------------------------------------------------------------------//
 //---------------------------------------------------------------------------//
+
+PRAGMA_PUSH_DEP_DECL
+
 static PyTypeObject PyRelay_Web_WebServer_TYPE = {
    PyVarObject_HEAD_INIT(NULL, 0)
    "WebServer",
@@ -590,6 +599,7 @@ static PyTypeObject PyRelay_Web_WebServer_TYPE = {
    PyVarObject_TAIL
 };
 
+PRAGMA_POP_DEP_DECL
 
 //---------------------------------------------------------------------------//
 // Leave commented until we need to use.
@@ -732,6 +742,9 @@ static PyMethodDef PyRelay_Web_WebSocket_METHODS[] = {
 
 
 //---------------------------------------------------------------------------//
+
+PRAGMA_PUSH_DEP_DECL
+
 static PyTypeObject PyRelay_Web_WebSocket_TYPE = {
    PyVarObject_HEAD_INIT(NULL, 0)
    "WebSocket",
@@ -782,6 +795,8 @@ static PyTypeObject PyRelay_Web_WebSocket_TYPE = {
    0
    PyVarObject_TAIL
 };
+
+PRAGMA_POP_DEP_DECL
 
 //---------------------------------------------------------------------------//
 // Leave commented until we need to use.

--- a/src/tests/conduit/t_conduit_json.cpp
+++ b/src/tests/conduit/t_conduit_json.cpp
@@ -709,7 +709,7 @@ TEST(conduit_json, json_parse_error_detailed)
         Node n_res;
         g.walk(n_res);
     }
-    catch(conduit::Error e)
+    catch(conduit::Error &e)
     {
         CONDUIT_INFO(e.message());
     }
@@ -723,7 +723,7 @@ TEST(conduit_json, json_parse_error_detailed)
         Node n_res;
         g.walk(n_res);
     }
-    catch(conduit::Error e)
+    catch(conduit::Error &e)
     {
         CONDUIT_INFO(e.message());
     }
@@ -736,7 +736,7 @@ TEST(conduit_json, json_parse_error_detailed)
         Node n_res;
         g.walk(n_res);
     }
-    catch(conduit::Error e)
+    catch(conduit::Error &e)
     {
         CONDUIT_INFO(e.message());
     }

--- a/src/tests/conduit/t_conduit_utils.cpp
+++ b/src/tests/conduit/t_conduit_utils.cpp
@@ -113,7 +113,7 @@ TEST(conduit_utils, error_constructors)
     {
         utils::handle_warning("ERROR!",__FILE__,__LINE__);
     }
-    catch(conduit::Error e)
+    catch(conduit::Error &e)
     {
         conduit::Error ecpy(e);
         CONDUIT_INFO(ecpy.message());
@@ -123,7 +123,7 @@ TEST(conduit_utils, error_constructors)
     {
         utils::handle_warning("ERROR!",__FILE__,__LINE__);
     }
-    catch(conduit::Error e)
+    catch(conduit::Error &e)
     {
         conduit::Error ecpy;
         ecpy = e;

--- a/src/tests/conduit/t_conduit_yaml.cpp
+++ b/src/tests/conduit/t_conduit_yaml.cpp
@@ -484,7 +484,7 @@ TEST(conduit_yaml, parse_error_detailed)
             Node n_res;
             g.walk(n_res);
         }
-        catch(conduit::Error e)
+        catch(conduit::Error &e)
         {
             CONDUIT_INFO(e.message());
              throw_occured = true;
@@ -503,7 +503,7 @@ TEST(conduit_yaml, parse_error_detailed)
             Node n_res;
             g.walk(n_res);
         }
-        catch(conduit::Error e)
+        catch(conduit::Error &e)
         {
             CONDUIT_INFO(e.message());
             throw_occured = true;
@@ -523,7 +523,7 @@ TEST(conduit_yaml, parse_error_detailed)
             Node n_res;
             g.walk(n_res);
         }
-        catch(conduit::Error e)
+        catch(conduit::Error &e)
         {
             CONDUIT_INFO(e.message());
             throw_occured = true;
@@ -543,7 +543,7 @@ TEST(conduit_yaml, parse_error_detailed)
             Node n_res;
             g.walk(n_res);
         }
-        catch(conduit::Error e)
+        catch(conduit::Error &e)
         {
             CONDUIT_INFO(e.message());
             throw_occured = true;

--- a/src/tests/docs/t_conduit_docs_tutorial_basics.cpp
+++ b/src/tests/docs/t_conduit_docs_tutorial_basics.cpp
@@ -242,7 +242,7 @@ TEST(conduit_tutorial, basics_const_vs_non_const)
     {
         important_suprise(n1);
     }
-    catch(conduit::Error e)
+    catch(conduit::Error &e)
     {
         e.print();
     }
@@ -263,7 +263,7 @@ TEST(conduit_tutorial, basics_const_vs_non_const)
     {
         important(n2);
     }
-    catch(conduit::Error e)
+    catch(conduit::Error &e)
     {
         e.print();
     }

--- a/src/tests/docs/t_conduit_docs_tutorial_errors.cpp
+++ b/src/tests/docs/t_conduit_docs_tutorial_errors.cpp
@@ -88,7 +88,7 @@ TEST(conduit_tutorial, error_handlers)
         const Node &n_my_value = n["my_value"];
         n_my_value["bad"];
     }
-    catch(conduit::Error e)
+    catch(conduit::Error &e)
     {
         // pass
     }

